### PR TITLE
feat: add support for migrations

### LINKED ISSUE #68: Alembic migrations trial
Our default schema should be managed in a more sane way. This ticket is to try to move the current schema to Alembic and see if there are any issues. If not, we should merge it and adopt that as our path forward. If there are, we should create a new ticket to adopt Tutor-style migrations.

The big concern here will be Tutor's jinja templating and whether we can make it play nice, especially with things like table and database name changes that Alembic will take for granted remaining the same. We should at least document the process of doing those things to prevent accidental data loss or other problems.

### PR DESCRIPTION
**Pull request by Ian2012**
## Description

This PR adds support for clickhouse migrations with Alembic and adds a command that serve as an interface to alembic.

In a tutor environment:
- To generate a migration run: `tutor local do alembic -c "revision -m 'my_migration'"`, and copy the file back to the versions folder in the plugin, but we encourage to just copy the last migration file, renamed with the next sequential ID, update the revision and down_revision variables.
- To upgrade to the latest migration run: `tutor local do alembic -c "upgrade head`
- To undo all migrations run: `tutor local do alembic -c "downgrade base`

Have in mind that the downgrade process will remove all tables and its data.

Fixes #68

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @Ian2012!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


This is currently a draft pull request. When it is ready for our review and all tests are green, click "Ready for Review", or remove "WIP" from the title, as appropriate.

File: tutoraspects/templates/aspects/jobs/init/aspects/init-aspects.sh [hunk 1]
```diff
@@ -3,6 +3,13 @@
 python3 -m venv virtualenv
 . virtualenv/bin/activate
 
+rm -rf /app/aspects/migrations/aspects-dbt
+
+echo "Run Alembic migrations..."
+cd /app/aspects/migrations
+
+alembic upgrade head
```
Comment (pomegranited) [inline tutoraspects/templates/aspects/jobs/init/aspects/init-aspects.sh:11]: Build failed -- need to install alembic before this step is run.

Comment (pomegranited) [reply tutoraspects/templates/aspects/jobs/init/aspects/init-aspects.sh:11]: Ah.. I see where they'll be installed on the image, so never mind :)

**Review by pomegranited — APPROVED** [review]
@Ian2012 This is working well (once I figured out I needed to rebuild my aspects image 😄 )

But as noted, I think it could use some more documentation before it merges.

Looking good! :+1:

- [x] I tested this on my tutor local
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [ ] Includes documentation -- requested more

File: tutoraspects/templates/aspects/apps/aspects/migrations/README [hunk 2]
```diff
@@ -0,0 +1 @@
+clickhouse-client --query 'DROP DATABASE IF EXISTS alembic' && clickhouse-client --allow_deprecated_database_ordinary 1 --query 'CREATE DATABASE alembic Engine = Ordinary' && alembic upgrade head
```
[the remark this answers is not in the document]
Comment (Ian2012) [reply tutoraspects/templates/aspects/apps/aspects/migrations/README:1]: Yes, I will. Good place for this information

**Comment by Ian2012** [comment]
@bmtcril @pomegranited this is ready for review

**Comment by bmtcril** [comment]
Changes we talked about in chat:

- Database creation will be outside of Alembic
- Alembic versions table will be stored in the same database as the XAPI schema, so if it's changed somewhere, the history will work as it will assume no migration has been applied

**Comment by Ian2012** [comment]
@bmtcril @pomegranited This is ready for review. The new implementation 

 Here is the result, I think we should keep the database name as a setting, so we can perform some tests without changing our current test data and allow advanced use cases.

image
image

**Review by pomegranited — COMMENTED** [review]
Sorry @Ian2012 , I started testing this but ran out of time today.. here's some comments.

File: tutoraspects/templates/aspects/build/aspects/Dockerfile [hunk 3]
```diff
@@ -1,5 +1,17 @@
...
-RUN pip install dbt-core==1.4.0 dbt-clickhouse==1.4.1
+WORKDIR /app
+
+COPY ./requirements.txt /app/requirements.txt
+
+COPY *.sh .
+
+RUN chmod +x alembic.sh
+RUN chmod +x clickhouse-demo-xapi-data.sh
+RUN chmod +x dbt.sh
```
[the remark this answers is not in the document]
Comment (Ian2012) [reply tutoraspects/templates/aspects/build/aspects/Dockerfile:11]: No, I was trying to refactor that and forgot to remove the reference in the build process

File: tutoraspects/templates/aspects/apps/aspects/migrations/alembic/versions/0001_get_org_from_course_url.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    op.execute(
+        """
+        -- Create the terrible user defined function to parse the org out of course URLs
+        -- if we need to update this just add a COALESCE around the whole thing and put in
+        -- additional cases wrapped in nullIf's until we get them all. Other things we may find in these URLs eventually:
+        -- i4x://{org}/{rest of key}  Old Mongo usage keys
+        -- c4x://{org}/{rest of key}  Old Mongo assets
+        -- {org}/{rest of key} Old Mongo course keys
+        CREATE OR REPLACE FUNCTION get_org_from_course_url AS (course_url) ->
+        nullIf(EXTRACT(course_url, 'course-v1:([a-zA-Z0-9]*)'), '');
```
[the remark this answers is not in the document]
Comment (Ian2012) [reply tutoraspects/templates/aspects/apps/aspects/migrations/alembic/versions/0001_get_org_from_course_url.py:20]: Yep, I will create an issue to track this down

File: tutoraspects/templates/aspects/apps/aspects/migrations/README.rst [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+- Once a migration is generated, it is not modified.
+
+- Migrations must be generated in tutor dev or local environment and then copied
+  back to the repository.
+
+Next steps
+----------
+
+Further research is needed to implement Alembic branches. See 
+https://alembic.sqlalchemy.org/en/latest/branches.html for more information.
```
[the remark this answers is not in the document]
Comment (Ian2012) [reply tutoraspects/templates/aspects/apps/aspects/migrations/README.rst:57]: I think a better place would be this one https://github.com/openedx/openedx-aspects/tree/main/docs/decisions. I will proceed with the PR once I receive @bmtcril approval and document this there

File: tutoraspects/templates/aspects/apps/aspects/migrations/README.rst [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+Aspects interface for Alembic
+-----------------------------
+
+The Aspects project provides a custom interface for Alembic. This interface is
+available in the command line with the subcommand ``tutor local|dev alembic do -c "custom command"``.
+
+For example, to run the command ``alembic upgrade head`` in the dev environment, you would run::
+
+    tutor dev alembic do -c "upgrade head"
```
Comment (bmtcril) [inline tutoraspects/templates/aspects/apps/aspects/migrations/README.rst:17]: For all of these commands I had to do them in a different order: `tutor local do alembic -c "upgrade head"`

**Review by bmtcril — APPROVED** [review]
One comment on the docs, otherwise I think this is good to go!

**Comment by openedx-webhooks** [comment]
@Ian2012 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/tutor-contrib-aspects&entry.867055334=https://github.com/openedx/tutor-contrib-aspects/pull/99&entry.1484655318=https://github.com/Ian2012&entry.752974735=2023-06-27+17:16&entry.1917517419=2023-07-05+20:40&entry.2133058324=Yes) so we can improve your experience in the future.

**Pull request merged**